### ISSUE TITLE
Icons fail to show

### ISSUE THREAD

**Issue by zhaocai** [Pitmairen/selection-search#63]
From the `Inspect views → background/background.html` (Developer Mode), the following error message appears:

```log
background.html:1 Error in event handler: Error: Failed to execute 'toDataURL' on 'HTMLCanvasElement': Tainted canvases may not be exported.
    at IconLoader.getDataURL (chrome-extension://gipnlpdeieaidmmeaichnddnmjmcakoe/background/iconloader.js:40:40)
    at IconCollection.getAllIconURLs (chrome-extension://gipnlpdeieaidmmeaichnddnmjmcakoe/background/iconloader.js:141:31)
    at getIcons (chrome-extension://gipnlpdeieaidmmeaichnddnmjmcakoe/background/background.js:125:33)
    at chrome-extension://gipnlpdeieaidmmeaichnddnmjmcakoe/background/init.js:72:24
```

**Comment by Pitmairen** [comment]
Do you have any custom defined icon urls on any search engines?

**Comment by zhaocai** [comment]
> Do you have any custom defined icon urls on any search engines?

Yes, a lot.

**Comment by Pitmairen** [comment]
Do you have an example of some of the icon urls that fails to load?

**Comment by zhaocai** [comment]
@Pitmairen All the icons fail to show including the built-in ones such as `COPY`.  It looks like this:

![](https://z-1258356203.cos.ap-chengdu.myqcloud.com/img/share/Selection-Search_2019-06-22_04-32-07.jpg)


Some of the search engines I use are listed as below.


|Name	|Search url	| Icon url		|
------------ | ------------ | -------------|
|Copy	|COPY		| (Use default) |
| 🅶oo🅶🅻e | http://google.com/search?q=%s |(Use default)
|History |chrome://history/?q=%s |https://z-1258356203.cos.ap-chengdu.myqcloud.com/img/share/history-hover.png
Antonym|http://www.synonym.com/antonym/%s|http://mycroftproject.com/updateos.php/id0/sna-ant.gif


P.S. Chrome `Version 76.0.3809.36 (Official Build) beta (64-bit)` in Windows 10.

**Comment by zhaocai** [comment]
Based on the DevTools screenshot as below, I suppose the icons are correctly loaded into `memory/disk cache`. I am not sure if it is related to this: [Allowing cross-origin use of images and canvas - HTML: Hypertext Markup Language | MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_enabled_image)
 
![](https://z-1258356203.cos.ap-chengdu.myqcloud.com/img/share/Selection-Search_2019-06-22_04-43-14.jpg)

**Comment by Pitmairen** [comment]
I will have to check with that version of Chrome, I'm currently on an older version. Something has probably changed in the newer versions, hopefully there is a workaround.

**Comment by Pitmairen** [comment]
I just tried with Chrome 76.0.3809.36. And It was working fine. But when I tried to add one of your custom urls (http://mycroftproject.com/updateos.php/id0/sna-ant.gif), I got the same error and all icons failed to load.

The error remained after removing the mycroftproject icon, I had to reload the extension or restart chrome for it to work again.

Could you try to remove the mycroftproject icon and restart chrome and see if that helps?

You can also try to use the https version of the mycroftproject url, this also seems to be working. https://mycroftproject.com/updateos.php/id0/sna-ant.gif

The http version is redirected to the https version, maybe it is this redirect that is causing problems.

**Comment by zhaocai** [comment]
> The http version is redirected to the https version, maybe it is this redirect that is causing problems.

Yes, It works after I remove/replace everything `http` icon urls. 

Thanks for the help 👍

**Comment by Pitmairen** [comment]
I don't think there is a problem with every http url. I testes with some other http urls and they worked fine. But i'm glad it worked :)

**Comment by Droyk** [comment]
I was having the same problem for the past 2 to 3 weeks didn't report it coz I was procrastinating. tried many things incognito mode, disable all of my extensions, even made a new profile same result can't see any icons. https://gfycat.com/serenebreakablelice 
Icons fail to show with the below messages popping up on the console!!

`(index):1 Error handling response: TypeError: Cannot read property 'icons' of undefined
    at chrome-extension://gipnlpdeieaidmmeaichnddnmjmcakoe/popup/init_script.js:103:37`

`Unchecked runtime.lastError: The message port closed before a response was received.`

So I tried @zhaocai solution, at first It didn't work. but then I correctly formatted all of search URL & Icon URL you see I mentioned 'Search URL' yes that's important they need to stay consistent except sometimes (yeah I don't know why?) and also a single formatting mistake in any of the columns will make the icons disappear again, so be sure to correctly format them one by one.    


How To Reproduce
---

@Pitmairen if you want correctly reproduce the problem just make a new profile in chrome (coz it's easy) install your extension go to the options page and add this URL `https://google.co.uk/favicon.ico` in the `Icon URL` column. save the settings & that's it. now select any word on any website & you won't be able to see any icons. icons most definitely would fail to show. now go to the options page & correctly format the URL `https://www.google.co.uk/favicon.ico` now disable & re-enable the extension (else it won't work.) & BAM!! you can now see the icons. 

**This wasn't a problem before, either you did something which made the URL formatting necessary or the chrome DEV team did something, either way, I am happy I can see my favIcons again.**

Chrome version Version 76.0.3809.100 (Official Build) (64-bit)

**Comment by Pitmairen** [comment]
I haven't made any changes to the icon loading. The issues seems to have started with Chrome 76.  I will try to improve the icon loading so that a single bad icon does not break all icons.

**Comment by Droyk** [comment]
So this is a URL formatting issue or icon loading issue.

**Comment by Pitmairen** [comment]
I think it's both. A badly formatted url can cause the code that loads the icons in the background to crash. This causes all icons to disappear. It also seems like there are some new security restrictions in Chrome 76 that causes some valid icon urls not to load properly. This also causes the icon loader to crash.

**Comment by Pitmairen** [comment]
I have released a new version 0.8.49 that should be available soon. It should hopefully fix the icon issue.

**Comment by Droyk** [comment]
Okay, that solved the problem!! I imported all of the badly formatted URLs again to see if they disappear and it seems like this update solved the issue, icons don't disappear anymore because of URL formatting.
